FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RecipientsViewModel.kt

```diff
@@ -0,0 +1,391 @@
+      if (emails.isEmpty()) {
+        return@launch
+      }
+
+      setResultForRemoteContactsLiveData(type, Result.loading())
+
+      val recipients = ArrayList<RecipientWithPubKeys>()
+      try {
+        for (email in emails) {
+          if (GeneralUtil.isEmailValid(email)) {
+            val emailLowerCase = email.lowercase(Locale.getDefault())
+            var cachedRecipientWithPubKeys = getCachedRecipientWithPubKeys(emailLowerCase)
+
+            if (cachedRecipientWithPubKeys == null) {
+              roomDatabase.recipientDao().insertSuspend(RecipientEntity(email = emailLowerCase))
+              cachedRecipientWithPubKeys =
+                roomDatabase.recipientDao().getRecipientWithPubKeysByEmailSuspend(emailLowerCase)
+            } else {
+              for (publicKeyEntity in cachedRecipientWithPubKeys.publicKeys) {
+                try {
+                  val result = PgpKey.parseKeys(publicKeyEntity.publicKey).pgpKeyDetailsList
+                  publicKeyEntity.pgpKeyDetails = result.firstOrNull()
+                } catch (e: Exception) {
+                  e.printStackTrace()
+                  publicKeyEntity.isNotUsable = true
+                }
+              }
+            }
+
+            try {
+              getPgpInfoFromServer(email = emailLowerCase)?.let { pgpKeyDetailsList ->
+                cachedRecipientWithPubKeys?.let { recipientWithPubKeys ->
+                  updateCachedInfoWithPubKeysFromLookUp(
+                    recipientWithPubKeys,
+                    pgpKeyDetailsList
+                  )
+                }
+              }
+              cachedRecipientWithPubKeys = getCachedRecipientWithPubKeys(emailLowerCase)
+              cachedRecipientWithPubKeys?.let { recipients.add(it) }
+            } catch (e: Exception) {
+              e.printStackTrace()
+              ExceptionUtil.handleError(e)
+            }
+          }
+        }
+        setResultForRemoteContactsLiveData(type, Result.success(recipients))
+      } catch (e: Exception) {
+        e.printStackTrace()
+        ExceptionUtil.handleError(e)
+        setResultForRemoteContactsLiveData(type, Result.exception(e))
+      }
+    }
+  }
+
+  private suspend fun getCachedRecipientWithPubKeys(emailLowerCase: String): RecipientWithPubKeys? =
+    withContext(Dispatchers.IO) {
+      val cachedRecipientWithPubKeys = roomDatabase.recipientDao()
+        .getRecipientWithPubKeysByEmailSuspend(emailLowerCase) ?: return@withContext null
+
+      for (publicKeyEntity in cachedRecipientWithPubKeys.publicKeys) {
+        try {
+          val result = PgpKey.parseKeys(publicKeyEntity.publicKey).pgpKeyDetailsList
+          publicKeyEntity.pgpKeyDetails = result.firstOrNull()
+        } catch (e: Exception) {
+          e.printStackTrace()
+          publicKeyEntity.isNotUsable = true
+        }
+      }
+      return@withContext cachedRecipientWithPubKeys
+    }
+
+  private suspend fun updateCachedInfoWithPubKeysFromLookUp(
```

Comment:
The logic in this method is correct but could use a lot of clarification to be easier to read first time. Please see comments below.